reduce dependency on ca65 charset translation behavior

mooinglemur opened the pull request:
Instead of relying on charset translations for string literals in the source, they should be taken directly as is.

A [recent commit](https://github.com/cc65/cc65/pull/2642) to cc65's repo appears to change the behavior of character literals in a non-backward compatible way, so that there is no good way to maintain the ability to get the same output with all versions of ca65 when dealing with character literals and math, unless we do away with conversions altogether.

mooinglemur committed:
reduce dependency on ca65 charset translation behavior

forgot to save files

stefan-b-jakobsson commented:
I currently use cc65 v 2.19 on MacOS installed with Homebrew.

"ascii_charmap.inc" is not to be found in the asminc folder. Make fails complaining about the missing file.

v 2.19 was released on Nov 20, 2020.

"ascii_charmap.inc" seems to have been added not until Apr 18, 2021.

After copying the content of that file to the basload-rom folder, make works.

I think we need to do something about this. Main alternatives:

1. Include ascii_charmap.inc in the basload source. The cc65 license is permissive, and I think it should be OK.

2. If the problem is only single chars plus an offset, could we get around the whole problem by replacing it with the resulting byte value? For instance, .byt "en", 'd' +$80 would become .byt "en", $c4.

mooinglemur commented:
In that case, I would remove the include, but then also remove the `-t cx16` from the build while still referencing the linker config, so that we forego PETSCII conversion and get ASCII encoding.  This is how `x16-rom` itself builds.

Simply replacing `'d' + $80` with `$C4` would still leave all of the other single quoted letters broken and assembling to the wrong value in the most recent builds of cc65.  The build errors served as a canary that there was a deeper problem.  Simply changing the ones that caused errors would only fix those, but we'd need to change all of the rest.

`cmp #'a'  ; with PETSCII conversion, $41 in old builds, and $C1 in new builds.`

mooinglemur commented:
I will update the PR tomorrow (Thursday).

stefan-b-jakobsson commented:
OK, I feel I need to look a bit deeper myself at how this exactly works.

A regular expression search returns two spots where there are still lower case letters in single quotes:

line.inc:774:
- This code looks for numbers in scientific notation, checking first for a lower case 'e' and then for an upper case 'E'. I tested it with upper case and lower case in both ISO and PETSCII modes. I couldn't get it to fail.

symbol.inc:255
- The code here checks if we have run out of short variable names. Should this be upper case 'Z' as well?

Overall, impressive work, that you so quickly could change all that in someone else's code!

stefan-b-jakobsson commented:
I've done some testing with 2.19 vs current master.

I can't see that the target switch affects the output. In 2.19 lower case 'a' becomes 0x41, and in current master it becomes 0xc1 no matter what.

stefan-b-jakobsson commented:
It looks like cl65 in current master defaults to C64 as target, if not specified.

This is set in src/cl65/main.c:1448.

I also patched the TgtTranslateInit function in src/common/tgttrans.c. Before it's called there is ASCII to ASCII (no change). Target is 9 (=C64). It copies the C64 charmap where ASCII 0x41 is translated to 0xc1, and where ASCII 0x61 is translated to 0x41.

I also note that the resulting code is different when I assemble with ca65 and link with ld65 in two steps even though I use the same params and configs.

stefan-b-jakobsson commented:
Yep, as far as I can tell, Target is initialized to TGT_NONE in src/common/target.c:229, and is never changed by ca65 unless you specify a target with the -t switch.

This behavior is obviously different from cl65. I think you can get cl65 to use TGT_NONE by specifying "-t none".

mooinglemur committed:
remove ascii_charmap.inc, switch to calling assembler with -t none and --cpu

mooinglemur commented:
Updated.  This should resolve all the issues mentioned above.  I think the scientific notation one should have been 'E' and 'E' + $80 to match PETSCII, and that's what I changed it to, but for some reason it now (still) works with the older and newest cc65 regardless of whether the source file is 'E' ($45), 'e' ($65), or shifted PETSCII 'E' ($C5).

I'm not sure if there's any sort of case-folding happening anywhere else in X16EDIT or BASLOAD, but... it all works in my testing.

mooinglemur commented on response.inc:
```diff
@@ -48,26 +48,26 @@ RESPONSE_TOO_MANY_NESTED_IFS    = 18
-msg_invalid_param: .byt "invalid parameter in ", 0
-msg_invalid_controlcode: .byt "invalid control code in ",0
-msg_invalid_symfile: .byt "symfile not allowed in ", 0
-msg_symfile_io_err: .byt "couldn not open symfile in ", 0
```
I did notice the typo when doing the uppercase conversions, but I think this is the only such change I did.

stefan-b-jakobsson commented:
I must have forgot about that when adding scientific notation, but as you say, case conversion happens in file_readln. You can leave it as is, I can clean it up later.
